fpga_interchange/arch.h

```diff
@@ -0,0 +1,1121 @@
+    delay_t getBoundingBoxCost(WireId src, WireId dst, int distance) const;
+    delay_t getDelayEpsilon() const { return 20; }
+    delay_t getRipupDelayPenalty() const { return 120; }
+    delay_t getWireRipupDelayPenalty(WireId wire) const;
+    float getDelayNS(delay_t v) const { return v * 0.001; }
+    DelayInfo getDelayFromNS(float ns) const
+    {
+        DelayInfo del;
+        del.delay = delay_t(ns * 1000);
+        return del;
+    }
+    uint32_t getDelayChecksum(delay_t v) const { return v; }
+    bool getBudgetOverride(const NetInfo *net_info, const PortRef &sink, delay_t &budget) const;
+
+    // -------------------------------------------------
+
+    bool pack();
+    bool place();
+    bool route();
+    // -------------------------------------------------
+
+    std::vector<GraphicElement> getDecalGraphics(DecalId decal) const;
+
+    DecalXY getBelDecal(BelId bel) const;
+    DecalXY getWireDecal(WireId wire) const;
+    DecalXY getPipDecal(PipId pip) const;
+    DecalXY getGroupDecal(GroupId group) const;
+
+    // -------------------------------------------------
+
+    // Get the delay through a cell from one port to another, returning false
+    // if no path exists. This only considers combinational delays, as required by the Arch API
+    bool getCellDelay(const CellInfo *cell, IdString fromPort, IdString toPort, DelayInfo &delay) const;
+    // Get the port class, also setting clockInfoCount to the number of TimingClockingInfos associated with a port
+    TimingPortClass getPortTimingClass(const CellInfo *cell, IdString port, int &clockInfoCount) const;
+    // Get the TimingClockingInfo of a port
+    TimingClockingInfo getPortClockingInfo(const CellInfo *cell, IdString port, int index) const;
+
+    // -------------------------------------------------
+
+    // Whether or not a given cell can be placed at a given Bel
+    // This is not intended for Bel type checks, but finer-grained constraints
+    // such as conflicting set/reset signals, etc
+    bool isValidBelForCell(CellInfo *cell, BelId bel) const {
+        // FIXME: Implement this
+        return true;
+    }
+
+    // Return true whether all Bels at a given location are valid
+    bool isBelLocationValid(BelId bel) const {
+        // FIXME: Implement this
+        return true;
+    }
+
+    IdString getBelTileType(BelId bel) const { return IdString(locInfo(bel).name); }
+
+    std::unordered_map<WireId, Loc> sink_locs, source_locs;
+    // -------------------------------------------------
+    void assignArchInfo() {}
+
+    // -------------------------------------------------
+
+    static const std::string defaultPlacer;
+    static const std::vector<std::string> availablePlacers;
+
+    static const std::string defaultRouter;
+    static const std::vector<std::string> availableRouters;
+
+    // -------------------------------------------------
+    template <typename Id> const TileTypeInfoPOD &locInfo(Id &id) const
```

Comment:
jfyi, and I fully appreciate that this was copied from nextpnr-xilinx, but going forward I'd like to move to a convention where the shared APIs like `getBels` are camelCase and arch-specific functions used internally like `locType` are using underscores (eg `loc_type`) instead to make the split a bit clearer.

Reply:
Have you considered just adding base virtual functions that can then be marked with "override" to clearly mark functions belong to the `archapi`, instead of a naming convention?  `BaseCtx` might need to be a template class then, but it would be cleaner in some ways.

Reply:
Unfortunately that doesn't work as-is because, for better or for worse, the return types are duck-typed and could be e.g. a reference to a vector or a custom range type depending on the arch.

Reply:
I guess you could have a _lot_ of template parameters but I think that would get quite unwieldy.

Reply:
So the way I solved this in VPR was to have the template parameter just be a place holder for a bunch of typedefs. Take a look: 

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/9aa339635b28b91e12ad736a2c9c9e556772937b/vpr/src/route/gen/rr_graph_uxsdcxx_interface.h#L54-L110

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/9aa339635b28b91e12ad736a2c9c9e556772937b/vpr/src/route/rr_graph_uxsdcxx_serializer.h#L213-L248

Reply:
The other advantage to this approach is it provides a namespace friendly register of types without populating the namespace.  It potentially means we can avoid the wacky include guards that are used right now, e.g.:

https://github.com/YosysHQ/nextpnr/blob/0d9790421699a22fc6a5962b41910c3b7d089353/generic/arch.h#L20-L24

https://github.com/YosysHQ/nextpnr/blob/0d9790421699a22fc6a5962b41910c3b7d089353/generic/archdefs.h#L20-L24

It would be really nice if `arch.h` and `archdefs.h` didn't use all the macro magic, and instead were "normal" headers.